feat: add support for Python 3.10

aucampia opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #374 🦕

NOTE: system test fail for alembic, but that is the same as is happening on main branch, see https://github.com/googleapis/python-bigquery-sqlalchemy/issues/382 and https://github.com/googleapis/python-bigquery-sqlalchemy/issues/373

tseaver commented on docs/conf.py:
```diff
@@ -113,6 +113,7 @@
     "**/.nox/**/*",
     "samples/AUTHORING_GUIDE.md",
     "samples/CONTRIBUTING.md",
+    "samples/snippets/README.md",
```
Rather than having a duplicate `README`, it would be better just to add the language to `samples/snippets/README.rst` (reformatting the ReST as needed).

tseaver commented on noxfile.py:
```diff
@@ -30,8 +30,8 @@
 # We're using two Python versions to test with sqlalchemy 1.3 and 1.4.
-SYSTEM_TEST_PYTHON_VERSIONS = ["3.8", "3.9"]
-UNIT_TEST_PYTHON_VERSIONS = ["3.6", "3.7", "3.8", "3.9"]
+SYSTEM_TEST_PYTHON_VERSIONS = ["3.8", "3.9", "3.10"]
```
We really don't need to run systests on all versions:  better would be to have the lower one and the upper one, e.g.:

```python
SYSTEM_TEST_PYTHON_VERSIONS = ["3.8", "3.10"]
```

aucampia replied:
Changed system tests to only run on 3.8 and 3.10

tseaver commented on noxfile.py:
```diff
@@ -100,7 +100,7 @@ def default(session):
     if session.python == "3.8":
         extras = "[tests,alembic]"
-    elif session.python == "3.9":
+    elif session.python in {"3.9", "3.10"}:
```
We don't necessarily to add the `geography` extra to more than one session.

aucampia replied:
Changed to `session.python == "3.10"`

tseaver commented on owlbot.py:
```diff
@@ -31,10 +31,11 @@
-    unit_test_python_versions=["3.6", "3.7", "3.8", "3.9"],
-    system_test_python_versions=["3.8", "3.9"],
+    unit_test_python_versions=["3.6", "3.7", "3.8", "3.9", "3.10"],
+    system_test_python_versions=["3.8", "3.9", "3.10"],
```
As above (`noxfile.py` is actually generated from `owlbot.py`).

aucampia replied:
Changed to only run on 3.8 and 3.10

tseaver commented on owlbot.py:
```diff
@@ -136,7 +137,7 @@ def compliance(session):
     if session.python == "3.8":
         extras = "[tests,alembic]"
-    elif session.python == "3.9":
+    elif session.python in {"3.9", "3.10"}:
```
Likewise, as above.

aucampia replied:
Changed to `session.python == "3.10"`

tseaver commented on owlbot.py:
```diff
@@ -31,10 +31,11 @@
 extras_by_python = {
     "3.8": ["tests", "alembic"],
     "3.9": ["tests", "geography"],
+    "3.10": ["tests", "geography"],
```
Unneeded.

aucampia replied:
Moved geography to `3.10`

tseaver commented on samples/snippets/requirements.in:
```diff
@@ -0,0 +1,15 @@
+GeoAlchemy2
+shapely
+packaging
+pytz
```
We prefer not to be generating these files:  we use the Whitesource Renovate bot to keep the hand-rolled versions up-to-date.

aucampia replied:
Can you trigger it for this repo? Part of the problem here is that the original `samples/snippets/requirements.txt` requires [Fiona](https://pypi.org/project/Fiona/)==1.8.20 and this version does not include support for python 3.10, though I also don't see why Fiona is in there to begin with.

aucampia replied:
If you want I can just commit the `requirements{.test,}.txt` files and remove `requirements{.test,}.in` files.

tswast replied:
> If you want I can just commit the requirements{.test,}.txt files and remove requirements{.test,}.in files

Please do.

> I also don't see why Fiona is in there to begin with.

I believe it came in as a transitive dependency of geopandas. Though, I don't see geopandas in use here: https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/samples/snippets/geography.py so perhaps leftover from an old sample.

aucampia replied:
Done

tseaver reviewed (commented):
Thanks very much for the patch!

aucampia committed:
feat: Enable support for 3.10

requirements{,-test}.txt are clean up to remove things that are not
needed for tests to pass and to include newer versions of dependencies.

aucampia commented:
Re-checked nox in `samples/snippets`:

```console
====================================================================== 2 passed, 6 warnings in 30.16s =======================================================================
nox > Session py-3.10 was successful.
nox > Running session readmegen
nox > Session readmegen skipped: This session had no parameters available..
nox > Ran multiple sessions:
nox > * lint: success
nox > * blacken: success
nox > * py-3.6: success
nox > * py-3.7: success
nox > * py-3.8: success
nox > * py-3.9: success
nox > * py-3.10: success
nox > * readmegen: skipped
20211229T193658 iwana@iwana-pc00.coop.no:~/sw/d/github.com/iafork/python-bigquery-sqlalchemy/samples/snippets
$ 
```

tswast reviewed (approved):
Thanks for the contribution!

I'll work on getting the system tests passing so that we can merge this soon.

tswast committed:
Merge branch 'main' into iwana-20211221T1332-py310
